relay.go

```diff
@@ -258,52 +258,54 @@ func (r *Relay) handleHopStream(s inet.Stream, msg *pb.CircuitRelay) {
 	err = rd.ReadMsg(msg)
 	if err != nil {
 		log.Debugf("error reading stop response: %s", err.Error())
-		bs.Close()
+		bs.Reset()
 		r.handleError(s, pb.CircuitRelay_HOP_CANT_OPEN_DST_STREAM)
 		return
 	}
 
 	if msg.GetType() != pb.CircuitRelay_STATUS {
 		log.Debugf("unexpected relay stop response: not a status message (%d)", msg.GetType())
-		bs.Close()
+		bs.Reset()
 		r.handleError(s, pb.CircuitRelay_HOP_CANT_OPEN_DST_STREAM)
 		return
 	}
 
 	if msg.GetCode() != pb.CircuitRelay_SUCCESS {
 		log.Debugf("relay stop failure: %d", msg.GetCode())
-		bs.Close()
+		bs.Reset()
 		r.handleError(s, msg.GetCode())
 		return
 	}
 
 	err = r.writeResponse(s, pb.CircuitRelay_SUCCESS)
 	if err != nil {
 		log.Debugf("error writing relay response: %s", err.Error())
-		bs.Close()
-		s.Close()
+		bs.Reset()
+		s.Reset()
 		return
 	}
 
 	// relay connection
 	log.Infof("relaying connection between %s and %s", src.ID.Pretty(), dst.ID.Pretty())
 
+	// Don't reset streams after finishing or the other side will get an
+	// error, not an EOF.
 	go func() {
 		count, err := io.Copy(s, bs)
 		if err != io.EOF && err != nil {
 			log.Debugf("relay copy error: %s", err)
 		}
-		s.Close()
```

Comment:
Small nitpick -- is there a reason for the reordering?
I'd like to have all the i/o related functionality grouped together, before the log message.

Reply:
No reason, I originally moved this to a defer but then decided against it and put it back in the wrong place. I'll put it back.

Reply:
ok, it does read nicer after the reordering!